### PR TITLE
SEARCH-7

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

backport the fix from iresearch upstream

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Review by KVS85 — APPROVED** [review]
LGTM.

File: 3rdParty/iresearch/core/search/bm25.cpp [hunk 1]
```diff
@@ -462,7 +462,7 @@ class sort final : public irs::prepared_sort_basic<bm25::score_t, bm25::stats> {
               irs::sort::score_cast<score_t>(state.score_buf) = state.filter_boost_->value *
                                                                 state.num_ *
                                                                 tf /
-                                                                (state.norm_const_ + state.norm_length_ * state.norm_.read() + tf);
+                                                                (state.norm_const_ + state.norm_length_ / state.norm_.read() + tf);
```
Comment (jsteemann) [inline 3rdParty/iresearch/core/search/bm25.cpp:465]: Is it possible that `state.norm_.read()` is 0 here in any case?

File: 3rdParty/iresearch/core/search/bm25.cpp [hunk 2]
```diff
@@ -473,7 +473,7 @@ class sort final : public irs::prepared_sort_basic<bm25::score_t, bm25::stats> {
               auto& state = *static_cast<bm25::norm_score_ctx*>(ctx);
 
               const float_t tf = ::SQRT(state.freq_->value);
-              irs::sort::score_cast<score_t>(state.score_buf) = state.num_ * tf / (state.norm_const_ + state.norm_length_ * state.norm_.read() + tf);
+              irs::sort::score_cast<score_t>(state.score_buf) = state.num_ * tf / (state.norm_const_ + state.norm_length_ / state.norm_.read() + tf);
```
Comment (jsteemann) [inline 3rdParty/iresearch/core/search/bm25.cpp:476]: Same here, is it possible that `state.norm_.read()` is 0 here in any case?

File: 3rdParty/iresearch/core/search/disjunction.hpp [hunk 3]
```diff
@@ -1388,8 +1390,8 @@ class block_disjunction final
   size_t match_count_;
   cost cost_;
   size_t buf_offset_{}; // offset within a buffer
-  score_buffer_type score_buf_;
-  min_match_buffer_type match_buf_;
+  score_buffer_type score_buf_;     // FIXME EBO
+  min_match_buffer_type match_buf_; // FIXME EBO
```
Comment (jsteemann) [inline 3rdParty/iresearch/core/search/disjunction.hpp:1394]: What does `EBO` mean in this context? I guess it refers to some optimization that we can do in the future?

File: 3rdParty/iresearch/core/search/bm25.cpp (hunk 1 above)
Comment (jsteemann) [reply 3rdParty/iresearch/core/search/bm25.cpp:465]: Checked with @gnusi that there are no 0-length values in the index.

File: 3rdParty/iresearch/core/search/bm25.cpp (hunk 2 above)
Comment (jsteemann) [reply 3rdParty/iresearch/core/search/bm25.cpp:476]: Checked with @gnusi that there are no 0-length values in the index.

File: 3rdParty/iresearch/core/search/disjunction.hpp (hunk 3 above)
Comment (jsteemann) [reply 3rdParty/iresearch/core/search/disjunction.hpp:1394]: "empty base optimization"

**Review by jsteemann — APPROVED** [review]
LGTM!

**Pull request merged**